### PR TITLE
Fixing null reference in propertyfetcher

### LINKED ISSUE #1349: NullReferenceException in HttpInListener
# Bug Report

OpenTelemetry.Extensions.Hosting" Version="0.6.0-beta.1"
OpenTelemetry.Instrumentation.AspNetCore" Version="0.6.0-beta.1"
OpenTelemetry.Instrumentation.Http" Version="0.6.0-beta.1" 

Runtime version netcoreapp3.1

## Symptom

When calling a site in our Asp.Net core application a NullReferenceException ocurrs within the HttpInListener (see screenshot)
error_httpInListener
)

Callstack:
image

Repro project is attached.
[MVCAppWithTracing.zip](https://github.com/open-telemetry/opentelemetry-dotnet/files/5379084/MVCAppWithTracing.zip)

The call is executed without problems, I guess the exception will be catched somewhere.

### PR DESCRIPTION
**Pull request by eddynaka**
Fixes #1349.

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry/Instrumentation/PropertyFetcher.cs [hunk 1]
```diff
@@ -45,6 +45,11 @@ public PropertyFetcher(string propertyName)
         /// <returns>Property fetched.</returns>
         public T Fetch(object obj)
         {
+            if (obj == null)
+            {
+                return default;
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: I feel like this should be a throw. Caller should check if the object is null before calling. PropertyFetcher is essentially analogous to `obj.Property` which would throw if obj is null. Returning default could be surprising and requires a post-check. @reyang @cijothomas Any thoughts?

Comment (eddynaka) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: i was in doubt when i did that as well. the only problem that each time we will have to verify the object before use.

Comment (CodeBlanch) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: How about... `Fetch` throws if obj is null and we add a `bool TryFetch(object obj, out T value)` for this case?

Comment (ejsmith) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: Can we have that return false if the wrong object type (not the first one) is passed into it?

Comment (eddynaka) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: @eksmith, what is your idea? What would be the difference from TryFetch?

Comment (eddynaka) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: Following @CodeBlanch  idea, i was planning to do this:
```csharp
public bool TryFetch(object obj, out T value)
        {
            if (obj == null)
            {
                value = default;
                return false;
            }

            value = Fetch(obj);
            return true;
        }
```

Comment (ejsmith) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: The `TypedPropertyFetch` gets passed the wrong wrong incompatible object type and just returns default. That behavior is super misleading and caused me great confusion. So I'm saying that if you add a `TryFetch` that returns a bool, which I think is great, then I'd like it to return false meaning it was unable to get the value.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/f14838c8de5c78fb29051bcf7990ded516304b1f/src/OpenTelemetry/Instrumentation/PropertyFetcher.cs#L106

Comment (eddynaka) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: i can check in that if the value is default, if yes, so we couldnt get the value....but that would be an assumption...what do u think?

Comment (ejsmith) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: Yes, that is a very bad assumption. I would think we would change the `TypedPropertyFetch` to use the `TryGet` pattern so we know exactly what happened.

Comment (eddynaka) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:50]: @ejsmith, not sure how that works exactly. at least the property doesnt contain the trygetmethod. would like to take a look? or should we do 2 parts: 1 with the `tryfetch` and other propagating with `tryget`?

**Review by cijothomas — APPROVED** [review]

File: src/OpenTelemetry/Instrumentation/PropertyFetcher.cs [hunk 2]
```diff
@@ -39,12 +39,19 @@ public PropertyFetcher(string propertyName)
         }
 
         /// <summary>
-        /// Fetch the property from the object.
+        /// Try to fetch the property from the object.
         /// </summary>
         /// <param name="obj">Object to be fetched.</param>
-        /// <returns>Property fetched.</returns>
-        public T Fetch(object obj)
```
Comment (ejsmith) [inline src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:46]: Hmm... I was thinking that this would stay for usages where you don't care and the TryFetch would be added. But I guess it's bad practice to not know if the operation succeeded or not. Just makes the call site for all the usages of this a little uglier.

Comment (eddynaka) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:46]: for users who doesn't care if that works or not, they will just use it and use the out variable. But, we shouldn't throw errors to the user. so, that's the overall idea :)

Comment (ejsmith) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:46]: Agreed. Good call. :-)

Comment (CodeBlanch) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:46]: If we want to keep it around, we could do:
```csharp
        public T Fetch(object obj)
        {
            if (!this.TryFetch(obj, out T value))
            {
                throw new ArgumentException("Supplied object was null or did not match the expected type.", nameof(obj));
            }

            return value;
        }
```

The idea was to remove the old one because it had hidden behaviors that could lead to errors. This version basically makes the hidden behaviors explicit.

Comment (eddynaka) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:46]: should i add or keep just with tryfetch?

Comment (ejsmith) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:46]: @CodeBlanch I like that. I think, like you've said before, this is working for most use cases without issues. So I kind of like the simpler call sites.

Comment (eddynaka) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:46]: let me add then 👍

Comment (eddynaka) [reply src/OpenTelemetry/Instrumentation/PropertyFetcher.cs:46]: just added!

File: src/OpenTelemetry.Instrumentation.GrpcNetClient/Implementation/GrpcClientDiagnosticListener.cs [hunk 3]
```diff
@@ -42,7 +42,7 @@ public GrpcClientDiagnosticListener(ActivitySourceAdapter activitySource, GrpcCl
 
         public override void OnStartActivity(Activity activity, object payload)
         {
-            if (!(this.startRequestFetcher.Fetch(payload) is HttpRequestMessage request))
+            if (!this.startRequestFetcher.TryFetch(payload, out HttpRequestMessage request))
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.GrpcNetClient/Implementation/GrpcClientDiagnosticListener.cs:45]: I think this should be `if (!this.startRequestFetcher.TryFetch(payload, out HttpRequestMessage request) || request == null)` to be equivalent to what it was before. Same comment on a few of them below this.

File: test/OpenTelemetry.Tests/Instrumentation/PropertyFetcherTest.cs [hunk 4]
```diff
@@ -37,10 +36,10 @@ public void FetchInvalidProperty()
         {
             var activity = new Activity("test");
             var fetch = new PropertyFetcher<string>("DisplayName2");
-            var result = fetch.Fetch(activity);
+            Assert.False(fetch.TryFetch(activity, out string result));
```
Comment (CodeBlanch) [inline test/OpenTelemetry.Tests/Instrumentation/PropertyFetcherTest.cs:39]: Would you add a test case for the null (eg: `fetch.TryFetch(null, out string result)`)? I forgot to do that in my branch sorry!

File: src/OpenTelemetry.Instrumentation.GrpcNetClient/Implementation/GrpcClientDiagnosticListener.cs (hunk 3 above)
Comment (eddynaka) [reply src/OpenTelemetry.Instrumentation.GrpcNetClient/Implementation/GrpcClientDiagnosticListener.cs]: just updated! Can you check if i missed anything?

File: test/OpenTelemetry.Tests/Instrumentation/PropertyFetcherTest.cs (hunk 4 above)
Comment (eddynaka) [reply test/OpenTelemetry.Tests/Instrumentation/PropertyFetcherTest.cs:39]: done

**Review by CodeBlanch — APPROVED** [review]
LGTM
